vulture_os/toolkit/api_parser/infoblox_threat_defense/infoblox_threat_defense.py

```diff
@@ -0,0 +1,148 @@
+            return True
+
+        except Exception as err:
+            raise InfobloxThreatDefenseAPIError(err)
+
+    def execute_query(self, url, params={}, timeout=20):
+
+        response = self.session.get(
+            url,
+            params=params,
+            proxies=self.proxies,
+            timeout=timeout,
+            verify=self.api_parser_custom_certificate or self.api_parser_verify_ssl
+        )
+        if response.status_code != 200:
+            error = f"Error at Infoblox_Threat_Defense API Call: status : {response.status_code}, content : {response.content}"
+            logger.error(f"[{__parser__}]:execute_query: {error}", extra={'frontend': str(self.frontend)})
+            raise InfobloxThreatDefenseAPIError(error)
+
+        return response.json()
+
+    def get_logs(self, since, to, offset=0):
+        self._connect()
+        url = f"{self.infoblox_threat_defense_host}/api/dnsdata/v2/dns_event"
+
+        params = {
+            "t0": int(since.timestamp()),
+            "t1": int(to.timestamp()),
+            "_offset": offset,
+            "_limit": 5000
+        }
+        response = self.execute_query(url, params=params)
+
+        if response["status_code"] == "200" and response["result"]:
+            return response["result"]
+        elif response["status_code"] != "200":
+            error = f"Error at Infoblox_Threat_Defense API Call: API returns a {response.get('status_code')} status code in its JSON content"
+            logger.error(f"[{__parser__}]:get_logs: {error}", extra={'frontend': str(self.frontend)})
+            raise InfobloxThreatDefenseAPIError(error)
+
+
+    def test(self):
+        try:
+            logs = self.get_logs(timezone.now()-timedelta(days=1), timezone.now())
+            return {
+                "status": True,
+                "data": logs
+            }
+        except Exception as e:
+            logger.exception(f"[{__parser__}]:test: {e}", extra={'frontend': str(self.frontend)})
+            return {
+                "status": False,
+                "error": str(e)
+            }
+
+    def format_log(self, log):
+        return json_dumps(log)
+
+    def execute(self):
+        since = self.last_api_call or (timezone.now() - timedelta(days=30))
+        to = min(timezone.now(), since + timedelta(hours=24))
+
+        msg = f"Parser starting from {since} to {to}"
+        logger.info(f"[{__parser__}]:execute: {msg}", extra={'frontend': str(self.frontend)})
+
+        offset = 0
+        while not self.evt_stop.is_set():
+
+            logs = self.get_logs(since, to, offset)
+
+            if not logs:
+                break
+
+            offset += len(logs)
+
+            self.write_to_file([self.format_log(l) for l in logs])
+
+            # Writting may take some while, so refresh token in Redis
+            self.update_lock()
+
+        # increment by 1s to avoid duplication of logs
+        self.frontend.last_api_call = to + timedelta(seconds=1)
```

Comment:
be careful, as this is the drawback of taking the `self.evt_stop` into account: you ought to handle the partial collection of logs in the `since:to` range when that happens!

Reply:
It's better than loss some logs in case of replay (we doesn't stock the offset), isn't it ?

Reply:
I was refering to taking the last timestamp of processed logs, and replacing the last_api_call value with it for the next iteration